# ExSenescence
Analysis codes and documentations for "Unmasking the heterogeneities in transcriptomic and evolutionary trajectories of cell cycle re-engaging post-mitotic neurons".     

## Data and codes used to identify Senescence neurons 
|Script name|Comment| 
|-----------|-------| 
|[1.CellCycleScore.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/SenescenceNeuronIdentification/1.CellCycleScore.code.R)|This script was used to quantify the cell cycle score accross the subtypes|
|[2.InferCNV.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/SenescenceNeuronIdentification/2.InferCNV.code.R)|This script was used to examine the copy number variation accross the subtype|
|[3.SenescenceSignature.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/SenescenceNeuronIdentification/3.SenescenceSignature.code.R)|This script was used to validate the senescence associated signatures of target subtypes|
|[SenescenceNeuronIdentification/ResultGraph/](https://github.com/KimChow-Lab/ExSenescence/tree/main/SenescenceNeuronIdentification/ResultGraph)|This file list the tmp results from the above scripts|


## Data and codes used to generate the figures  
|Figure name|Comment|The data underlying this figure can be found at metaData file |The codes underlying this figure can be found at scripts file|
|-----------|-------|-----------------------------------------------|------------------------------------------------|    
| Figure 1 | Identification of cell cycle re-engaging post-mitotic excitatory neurons | [Metadata_Figure 1](https://github.com/KimChow-Lab/ExSenescence/blob/main/metaData/) | [InitialAnalysisUsingMathysnRNA.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/InitialAnalysisUsingMathysnRNA.code.R)<br>[InferCNV.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/InferCNV.code.R)
| Figure 2 | Validation analyses in multiple cohorts of datasets  | [Metadata_Figure 2](https://github.com/KimChow-Lab/ExSenescence/blob/main/metaData/) | [snRNAPreprocessing.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/snRNAPreprocessing.code.R)<br>[AllExIntegrated.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/AllExIntegrated.code.R)<br>[InferCNV.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/InferCNV.code.R)
| Figure 3 | Senescence as the immediate terminal fate of cell cycle gene re-expressing neurons | [Metadata_Figure 3](https://github.com/KimChow-Lab/ExSenescence/blob/main/metaData/) | [AllExIntegrated.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/AllExIntegrated.code.R)<br>[monocleAnalysis.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/monocleAnalysis.code.R)
| Figure 4 | Early senescent neurons of AD brains exhibit unique pro-neuropathological signature | [Metadata_Figure 4](https://github.com/KimChow-Lab/ExSenescence/blob/main/metaData/) | [InferCNV.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/InferCNV.code.R)<br>[cellNumberVariation.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/cellNumberVariation.code.R)<br>[cellCommunication.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/cellCommunication.code.R)
| Figure S1 | Heatmap of CNV profile | [Metadata_Figure S1](https://github.com/KimChow-Lab/ExSenescence/blob/main/metaData/) | [InferCNV.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/InferCNV.code.R)
| Figure S2 | Brain cells with reserved mitotic capacities revealed limited signs of predicted CNVs | [Metadata_Figure S2](https://github.com/KimChow-Lab/ExSenescence/blob/main/metaData/) | [InferCNV.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/InferCNV.code.R)<br>[InitialAnalysisUsingMathysnRNA](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/InitialAnalysisUsingMathysnRNA.code.R)
| Figure S3 | Definition of excitatory neuronal sub-clusters was not interfered by methods | [Metadata_Figure S3](https://github.com/KimChow-Lab/ExSenescence/blob/main/metaData/) | [RobustParameter.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/RobustParameter.code.R)
| Figure S4 | Distribution of neuronal cells based on various conditions | [Metadata_Figure S4](https://github.com/KimChow-Lab/ExSenescence/blob/main/metaData/) | [InitialAnalysisUsingMathysnRNA](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/InitialAnalysisUsingMathysnRNA.code.R)<br>[AllExIntegrated.code](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/AllExIntegrated.code.R)
| Figure S5 | Prediction and profiling of copy number variation separately in multiple datasets | [Metadata_Figure S5](https://github.com/KimChow-Lab/ExSenescence/blob/main/metaData/) | [InferCNV.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/InferCNV.code.R)
| Figure S6 | Validation of cortical layer markers by spatial transcriptomic analysis. | [Metadata_Figure S6](https://github.com/KimChow-Lab/ExSenescence/blob/main/metaData/) | [spatialLIBD.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/spatialLIBD.code.R)
| Figure S7 | Single-cell trajectory analysis using the Monocle 2.0 | [Metadata_Figure S7](https://github.com/KimChow-Lab/ExSenescence/blob/main/metaData/) | [monocleAnalysis.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/monocleAnalysis.code.R)
| Figure S8 | Compromised DNA damage responses network in late senescent neurons | [Metadata_Figure S8](https://github.com/KimChow-Lab/ExSenescence/blob/main/metaData/) | [AllExIntegrated.code](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/AllExIntegrated.code.R)
| Figure S9 | Additional marker signatures of late senescent neurons. | [Metadata_Figure S9](https://github.com/KimChow-Lab/ExSenescence/blob/main/metaData/) | [AllExIntegrated.code](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/AllExIntegrated.code.R)
| Figure S10 | Impact of predicted CNVs on transcriptomic signature in sub-cluster 5 | [Metadata_Figure S10](https://github.com/KimChow-Lab/ExSenescence/blob/main/metaData/) | [AllExIntegrated.code](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/AllExIntegrated.code.R)
| Figure S11 | Characterization of upstream eventsof terminally senescent neurons | [Metadata_Figure S11](https://github.com/KimChow-Lab/ExSenescence/blob/main/metaData/) | [transcriptionFactorAnalysis.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/transcriptionFactorAnalysis.code.R)
| Figure S12 | AD risk gene loci mapping to predicted CNV locations | [Metadata_Figure S12](https://github.com/KimChow-Lab/ExSenescence/blob/main/metaData/) | [InferCNV.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/InferCNV.code.R)
| Figure S13 | Normalized expression levels of canonical AD-related CNV gene loci | [Metadata_Figure S13](https://github.com/KimChow-Lab/ExSenescence/blob/main/metaData/) | [AllExIntegrated.code](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/AllExIntegrated.code.R)
| Figure S14 | Senescent neurons exhibit compromised two-way communications with glia | [Metadata_Figure S14](https://github.com/KimChow-Lab/ExSenescence/blob/main/metaData/) | [cellCommunication.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/cellCommunication.code.R)
| Figure S15-19 | Correlation plots between the cell number ratios of different neuronal sub-clusters based on various status | [Metadata_Figure 15-19](https://github.com/KimChow-Lab/ExSenescence/blob/main/metaData/) | [cellNumberVariation.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/cellNumberVariation.code.R)
| Figure S20 | Application of the bioinformatics analytical pipeline in the Parkinsonâ€™s disease(PD) | [Metadata_Figure S16](https://github.com/KimChow-Lab/ExSenescence/blob/main/metaData/) | [Parkinson.DopamineSenescence.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/Parkinson.DopamineSenescence.code.R)
| Figure S21 | Supporting data for the analysis performed with the PD | [Metadata_Figure S17](https://github.com/KimChow-Lab/ExSenescence/blob/main/metaData/) | [Parkinson.DopamineSenescence.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/Parkinson.DopamineSenescence.code.R)
| Figure S22-23 | Application of the bioinformatics analytical pipeline in the mouse model of brain aging | [Metadata_Figure S22-23](https://github.com/KimChow-Lab/ExSenescence/blob/main/metaData/) | [Mouse_Allen_GSE207848.code.R](https://github.com/KimChow-Lab/ExSenescence/blob/main/scripts/Mouse_Allen_GSE207848.code.R)

## codes used to identify senescence neurons
